clickhouse/index.js

```diff
@@ -92,6 +93,119 @@ class ClickHouse {
+        Payload String CODEC(ZSTD(1)),
+
+        -- Bloom filter indices: pattern borrowed from OTel/logging schemas where
+        -- bloom filters are applied to frequently-filtered dimensions. Effectiveness
+        -- depends on correlation with ORDER BY and actual query patterns.
+        INDEX idx_orgid (OrgId) TYPE bloom_filter(0.01) GRANULARITY 1,
+        INDEX idx_deviceid (DeviceId) TYPE bloom_filter(0.01) GRANULARITY 1,
+        INDEX idx_scope (Scope) TYPE bloom_filter(0.01) GRANULARITY 1,
+        INDEX idx_capability (CapabilityName) TYPE bloom_filter(0.01) GRANULARITY 1
+    )
+    ENGINE = MergeTree()
+    -- Partition by day, matching the time-based TTL. This is the same pattern used
+    -- in observability/log schemas to make dropping expired data efficient.
+    PARTITION BY toYYYYMMDD(Timestamp)
+
+    -- Primary key defines the sparse index; putting OrgId and DeviceId first
+    -- optimizes for queries scoped by org/device, then narrowed by time.
+    PRIMARY KEY (OrgId, DeviceId, Timestamp)
+
+    -- ORDER BY controls on-disk sort and primary index layout. This order is
+    -- tuned for access patterns like "all messages for org+device over time",
+    -- similar to (ServiceName, Timestamp, ...) in OTel examples.
+    ORDER BY (OrgId, DeviceId, Timestamp)
+
+    -- Table-level TTL: automatically removes data 30 days after Timestamp.
+    -- This follows the same pattern as otel_logs/otel_traces TTL definitions.
+    TTL toDateTime(Timestamp) + toIntervalDay(30)
+
+    SETTINGS
+      -- Default granularity used in the observability examples; balances index
+      -- size and skipping efficiency.
+      index_granularity = 8192,
+
+      -- Recommended for TTL-based retention: drop whole parts when all rows
+      -- are expired, avoiding expensive row-level TTL mutations.
+      ttl_only_drop_parts = 1`;
+
+    try {
+      await this.client.exec({
+        query,
+        clickhouse_settings: {
+          wait_end_of_query: 1,
+        }
+      });
+      this._alreadyEnsuredHistoryTable = true;
+    } catch (error) {
+      console.error('Error executing query:', error.message);
+      console.debug('Query was:', query);
+      throw error;
+    }
+  }
+
+  /** Register an MQTT topic for storage in ClickHouse
+   * subscribes to the topic and stores incoming messages
+   * in a ClickHouse table.
+   * @param {Object} dataCache - DataCache instance to use for subscribing
+   * @param {string} topic - MQTT topic to register
+   */
+  async registerMqttTopicForStorage(dataCache, topic) {
+    await this.ensureMqttHistoryTable();
+    // Subscribe to the topic using subscribePath to get objects as-is (not flattened to leaves)
+    dataCache.subscribePath(topic, async (value, topic, matched, tags) => {
+      const timestamp = new Date();
+      // Remove leading empty string caused by topic starting with '/'
+      const topicParts = topic.replace(/^\//, '').split('/');
```

Comment:
Please don't do this. We already have topicToPath and it handles the `/` correctly -- this code does not.